source/extensions/filters/http/credential_injector/credential_injector_filter.cc

```diff
@@ -0,0 +1,153 @@
+#include "source/extensions/filters/http/credential_injector/credential_injector_filter.h"
+
+#include "source/common/http/headers.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace CredentialInjector {
+
+FilterConfig::FilterConfig(CredentialInjectorSharedPtr credential_injector, bool overwrite,
+                           bool allow_request_without_credential, const std::string& stats_prefix,
+                           Stats::Scope& scope)
+    : injector_(credential_injector), overwrite_(overwrite),
+      allow_request_without_credential_(allow_request_without_credential),
+      stats_(generateStats(stats_prefix + "credential_injector.", scope)) {}
+
+// Inject configured credential to the HTTP request header.
+// return true if successful
+bool FilterConfig::injectCredential(Http::RequestHeaderMap& headers) {
+  absl::Status status = injector_->inject(headers, overwrite_);
+
+  // If the credential already exists in the header and overwrite is false,
+  // increase the counter and continue processing the request.
+  if (absl::IsAlreadyExists(status)) {
+    ASSERT(!overwrite_); // overwrite should be false if AlreadyExists is returned
+    ENVOY_LOG(trace, "Credential already exists in the header");
+    stats_.already_exists_.inc();
+    // continue processing the request if credential already exists in the header
+    return true;
+  }
+
+  // If failed to inject the credential
+  if (!status.ok()) {
+    ENVOY_LOG(debug, "Failed to inject credential: {}", status.message());
+    stats_.failed_.inc();
+    return allow_request_without_credential_;
+  }
+
+  // If successfully injected the credential
+  ENVOY_LOG(debug, "Successfully injected credential");
+  stats_.injected_.inc();
+  return true;
+}
+
+CredentialInjectorFilter::CredentialInjectorFilter(FilterConfigSharedPtr config)
+    : config_(std::move(config)) {}
+
+Http::FilterHeadersStatus CredentialInjectorFilter::decodeHeaders(Http::RequestHeaderMap& headers,
+                                                                  bool) {
+  // Initiate the credential provider if not already done.
+  if (!credential_init_) {
+    // Save the pointer to the request headers for header manipulation based on credential provider
+    // response later.
+    request_headers_ = &headers;
+
+    in_flight_credential_request_ = config_->requestCredential(*this);
+
+    // If the callback is called immediately, continue decoding.
+    // We don't need to inject credential here because the credential has been injected in the
+    // onSuccess callback.
+    if (credential_init_) {
+      return Http::FilterHeadersStatus::Continue;
+    }
+
+    // pause while we await the credential provider to retrieve the credential, for example, an
+    // oauth2 credential provider may need to make a remote call to retrieve the credential.
+    stop_iteration_ = true;
+    return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+  }
```

Comment:
hmmm, I think there is a logic error? You see, at the start time, the `credential_init_` must be false and you will enter this branch, then, the following code after this branch will never be reached?

Is I missed something?

Reply:
The credential has been initialized when the filter receives the first request. Then the follow requests in this connection reuse the cached credential.

Reply:
I thought the filter instance is per connection, it turns out it's per stream, right?

Reply:
> I assumed filter instances are per connection, but they're actually per request. Is that correct?

Yeah, unique HTTP filter instance will be created for per request.

> We'll need a thread local cache to store the credentials.

I didn't get this point. the sds-based credentials have used the TLS (thread local storage) to ensure the thread safe.
And for other credentials like oauth2 seems needn't that?